### ISSUE TITLE
Incorrect sequence points encountered when debugging Roslyn

### ISSUE THREAD

**Issue by jcouv** [dotnet/roslyn#15851]
When debugging the test below, the debugger appears to go into the body of a conditional despite the condition being false.

Here’s the repro I have so far:
-	Have a  breakpoint inside `IsPossibleDeconstructionDesignation`
-	Debug the test below
-	Step out of `IsPossibleDeconstructionDesignation` and step forward once
-	Notice the function returns false, but the steps apparently go into conditional block as-if the condition was true
- Step forward once more
- Notice the active statement is now `CreateSuggestionModeItem` (but that method doesn't actually gets called)
- Step forward once more
- Notice that the `return null;` is reached after all

```C#
                else if (tree.IsPossibleDeconstructionDesignation(position, cancellationToken))
                {
                    return CreateSuggestionModeItem(CSharpFeaturesResources.designation_name,
                        CSharpFeaturesResources.Autoselect_disabled_due_to_possible_deconstruction_declaration); 
                   // this code is not executed (which is correct), but the debugger makes it look like it does
                } 
            }

            return null;
        }
```

```VB
        <WpfFact, Trait(Traits.Feature, Traits.Features.Completion)>
        Public Async Function TestIncompleteParenthesizedDeconstructionDeclaration() As Task
            Using state = TestState.CreateCSharpTestState(
                  <Document><![CDATA[
class Variable
{
    public void Foo()
    {
       (var a $$
    }
}]]></Document>)

                state.SendTypeChars(", va")
                Await state.AssertSelectedCompletionItem(displayText:="var", isHardSelected:=True)
                state.SendTypeChars(" ")
                Await state.AssertNoCompletionSession()

                state.SendTypeChars("a")
                Await state.AssertSelectedCompletionItem(displayText:="as", isSoftSelected:=True)
                state.SendTypeChars(")")
                Assert.Contains("(var a, var a)", state.GetLineTextFromCaretPosition(), StringComparison.Ordinal)
                Await state.AssertNoCompletionSession()
            End Using
        End Function
```

FYI @tmat Do you have any tips for troubleshooting this?
CC @AlekseyTs who is my witness ;-)

**Comment by gafter** [comment]
See also #12564

**Comment by jcouv** [comment]
Thanks @gafter for making the connection. This is very likely related. 

The method I encountered the problem is `async` as well:
```C#
  protected override async Task<CompletionItem> GetSuggestionModeItemAsync(
            Document document, int position, TextSpan itemSpan, CompletionTrigger trigger, CancellationToken cancellationToken = default(CancellationToken))
```

**Comment by jcouv** [comment]
Closing as duplicate of #12564

**Comment by jcouv** [comment]
Fixed by #16803 (Missing sequence point on implicit return in async method)
The fix unfortunately didn't make it into RTW, but will be released in the next ship window.
